src/porepy/numerics/ad/operator_functions.py

```diff
@@ -0,0 +1,463 @@
+
+
+### CONCRETE IMPLEMENTATIONS ------------------------------------------------------------------
+
+
+class Function(AbstractFunction):
+    """Ad representation of an analytically given function,
+    where it is expected that feeding the callable with Ad_arrays directly will give
+    give the proper result.
+
+    Here the values AND the Jacobian are obtained exactly by standard rules of differentiation.
+
+    The intended use is as a wrapper for operations on pp.ad.Ad_array objects,
+    in forms which are not directly or easily expressed by the rest of the Ad
+    framework.
+
+    NOTE: this is a special case where `get_values` and `get_jacobian` are formally implemented
+    but never used by the AD framework
+    (due to the assumed, direct evaluation of result using the callable).
+    """
+
+    def __init__(
+        self, func: Callable, name: str, is_vector_func: Optional[bool] = True
+    ):
+        super().__init__(func, name, is_vector_func, True)
+
+    def get_values(self, *args: "Ad_array") -> "np.ndarray":
+        result = self.func(*args)
+        return result.val
+
+    def get_jacobian(self, *args: "Ad_array") -> "np.ndarray":
+        result = self.func(*args)
+        return result.jac
+
+
+class LJacFunction(AbstractJacFunction):
+    """
+    Approximates the Jacobian of the function using the L-scheme
+    with a fixed value per dependency.
+    """
+
+    def __init__(
+        self,
+        L: Union[List[float], float],
+        func: Callable,
+        name: str,
+        is_vector_func: Optional[bool] = False,
+    ):
+        """Constructor.
+
+        The L-multiplier for the L-scheme can be passed for every argument specifically
+        using a list.
+        The order in the list has to mach the order of arguments when calling
+        this instance.
+
+        :param L: multiplier for identity for L-scheme
+        :type L: float / List[float]
+        """
+        super().__init__(func, name, is_vector_func)
+        # check and format input for further use
+        if isinstance(L, list):
+            self._L = [float(val) for val in L]
+        else:
+            self._L = [float(L)]
+
+    def get_jacobian(self, *args: "Ad_array") -> "sps.spmatrix":
+        """The approximate jacobian is identity times L.
+
+        Where the respective blocks appears,
+        depends on the total DOFs and the order of arguments passed during the
+        call to this instance.
+        """
+        # the Jacobian of a (Merged) Variable is already a properly sized block identity
+        if len(args) >= 1:
+            jac = args[0].jac * self._L[0]
+
+            # summing identity blocks for each dependency
+            if len(args) > 1:
+                # TODO think about exception handling in case not enough
+                # L-values were provided initially
+                for arg, L in zip(args[1:], self._L[1:]):
+                    jac += arg.jac * L
```

Comment:
I may understand how this is supposed to work, but I'm not sure - and I'm pretty sure a client will misunderstand. Also, be aware that for a problem with multiple variables, the full jac of arg (not only the block related to 'itself') - is this indended?

Reply:
I am not sure I understand properly
The Jacobian constructed here is exact for the following function:
L_1 * x_1 + L_2 * x_2 + L_3 * x_3 +....
where x_i is a variable passed during the function call (e.g. `density(pressure, temperature)` density -function, pressure and temp - variables)
The multipliers L_i are the multipliers passed at instantiation.